flink-runtime/src/main/java/org/apache/flink/runtime/state/ChangelogTaskLocalStateStore.java

```diff
@@ -0,0 +1,255 @@
+    protected File getCheckpointDirectory(long checkpointId) {
+        final File checkpointDirectory =
+                localRecoveryConfig
+                        .getLocalStateDirectoryProvider()
+                        .orElseThrow(
+                                () -> new IllegalStateException("Local recovery must be enabled."))
+                        .subtaskBaseDirectory(checkpointId);
+        File directoryForChangelog =
+                new File(checkpointDirectory, CHANGE_LOG_CHECKPOINT_PREFIX + checkpointId);
+
+        if (!directoryForChangelog.exists() && !directoryForChangelog.mkdirs()) {
+            throw new FlinkRuntimeException(
+                    String.format(
+                            "Could not create the checkpoint directory '%s'",
+                            directoryForChangelog));
+        }
+
+        return directoryForChangelog;
+    }
+
+    private void deleteMaterialization(LongPredicate pruningChecker) {
+        final Set<Long> materializationToRemove = new HashSet<>();
+        synchronized (lock) {
+            Iterator<Entry<Long, Long>> iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                long entryCheckpointId = entry.getKey();
+                if (pruningChecker.test(entryCheckpointId)) {
+                    materializationToRemove.add(entry.getValue());
+                    iterator.remove();
+                }
+            }
+
+            iterator = mapToMaterializationId.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Long, Long> entry = iterator.next();
+                materializationToRemove.remove(entry.getValue());
+            }
+        }
+
+        for (Long materializationId : materializationToRemove) {
+            File materializedDir =
+                    localRecoveryConfig
+                            .getLocalStateDirectoryProvider()
+                            .orElseThrow(
+                                    () ->
+                                            new IllegalStateException(
+                                                    "Local recovery must be enabled."))
+                            .subtaskSpecificCheckpointDirectory(materializationId);
+
+            if (!materializedDir.exists()) {
+                continue;
+            }
+            try {
+                deleteDirectory(materializedDir);
+            } catch (IOException ex) {
+                LOG.warn(
+                        "Exception while deleting local state directory of materialized part {} in subtask ({} - {} - {}).",
+                        materializedDir,
+                        jobID,
+                        jobVertexID,
+                        subtaskIndex,
+                        ex);
+            }
+        }
+    }
+
+    @Override
+    public void confirmCheckpoint(long confirmedCheckpointId) {
```

Comment:
Instead of overriding (and duplicating) these two methdos (confirm & abort)
can we override `pruneCheckpoints`?
Deleting materialization seems to be a part of pruning...

```
    @Override
    protected void pruneCheckpoints(LongPredicate pruningChecker, boolean breakOnceCheckerFalse) {
        deleteMaterialization(pruningChecker);
        super.pruneCheckpoints(pruningChecker, breakOnceCheckerFalse);
    }
```